pkg/webhook/v1beta1/experiment/validator/validator.go

```diff
@@ -420,79 +472,93 @@ func validatePatchJob(runSpec *unstructured.Unstructured, job interface{}, jobTy
 		}
 		if !strings.HasPrefix(mcSpec.Source.HttpGet.Path, "/") {
-			return fmt.Errorf(".spec.metricsCollectorSpec.source.httpGet.path is invalid for metrics collector kind: %v", mcKind)
+			allErrs = append(allErrs, field.Invalid(metricsSourcePath.Child("httpGet").Child("path"),
+				mcSpec.Source.HttpGet.Path, fmt.Sprintf(".spec.metricsCollectorSpec.source.httpGet.path is invalid for metrics collector kind: %v", mcKind)))
 		}
 	case commonapiv1beta1.CustomCollector:
 		if mcSpec.Collector.CustomCollector == nil {
-			return fmt.Errorf(".spec.metricsCollectorSpec.collector.customCollector is required for metrics collector kind: %v", mcKind)
+			allErrs = append(allErrs, field.Required(metricsCollectorPath.Child("collector").Child("customCollector"),
+				fmt.Sprintf(".spec.metricsCollectorSpec.collector.customCollector is required for metrics collector kind: %v", mcKind)))
 		}
-		if mcSpec.Source.FileSystemPath != nil {
+		if mcSpec.Source != nil && mcSpec.Source.FileSystemPath != nil {
 			if !filepath.IsAbs(mcSpec.Source.FileSystemPath.Path) || (mcSpec.Source.FileSystemPath.Kind != commonapiv1beta1.DirectoryKind &&
 				mcSpec.Source.FileSystemPath.Kind != commonapiv1beta1.FileKind) {
-				return fmt.Errorf(".spec.metricsCollectorSpec.source is invalid")
+				allErrs = append(allErrs, field.Invalid(metricsSourcePath.Child("fileSystemPath"),
+					"", ".spec.metricsCollectorSpec.source is invalid"))
 			}
 		}
 	default:
-		return fmt.Errorf("invalid metrics collector kind: %v", mcKind)
+		allErrs = append(allErrs, field.Invalid(metricsCollectorPath.Child("collector").Child("kind"),
+			mcKind, fmt.Sprintf("invalid metrics collector kind: %v", mcKind)))
 	}
 	if mcSpec.Source != nil && mcSpec.Source.Filter != nil && len(mcSpec.Source.Filter.MetricsFormat) > 0 {
 		// the filter regular expression must have two top subexpressions, the first matched one will be taken as metric name, the second one as metric value
 		mustTwoBracket, _ := regexp.Compile(`.*\(.*\).*\(.*\).*`)
 		for _, mFormat := range mcSpec.Source.Filter.MetricsFormat {
 			if _, err := regexp.Compile(mFormat); err != nil {
-				return fmt.Errorf("invalid %q in .spec.metricsCollectorSpec.source.filter: %v", mFormat, err)
+				allErrs = append(allErrs, field.Invalid(metricsSourcePath.Child("filter").Child("metricsFormat"),
+					mFormat, fmt.Sprintf("invalid %q in .spec.metricsCollectorSpec.source.filter: %v", mFormat, err)))
 			} else {
 				if !mustTwoBracket.MatchString(mFormat) {
-					return fmt.Errorf("invalid %q in .spec.metricsCollectorSpec.source.filter: two top subexpressions are required", mFormat)
+					allErrs = append(allErrs, field.Invalid(metricsSourcePath.Child("filter").Child("metricsFormat"),
+						mFormat, fmt.Sprintf("invalid %q in .spec.metricsCollectorSpec.source.filter: two top subexpressions are required", mFormat)))
```

Comment:
Do you also want to refactor error messages for `metricsCollectorSpec` given that we we pass the Child info in the error ?

Reply:
E.g. the error message might look as follows:
```
[spec.metricsCollectorSpec.source.fileSystemPath.format: Required value: format of metrics file is required]
```

Reply:
@andreyvelich done, sorry for my delay